pkg/applicationserver/distribution/distribution.go

```diff
@@ -0,0 +1,38 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package distribution
+
+import (
+	"context"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+// PubSub sends upstream traffic from publishers to subscribers.
+type PubSub interface {
+	// SendUp publishes traffic to the subscribers.
+	SendUp(context.Context, *ttnpb.ApplicationUp) error
```

Comment:
I would assume this is blocking because of the name and the fact it returns an error. So that's ok, but please do this concurrent then

Reply:
You might want to rename these to `Publish()` so that they match better with their `Subscribe()` counterparts and it aligns better with their godoc